src/wallet/scriptpubkeyman.cpp

```diff
@@ -839,10 +894,27 @@ bool LegacyScriptPubKeyMan::AddWatchOnly(const CScript& dest, int64_t nCreateTim
 void LegacyScriptPubKeyMan::SetHDChain(const CHDChain& chain, bool memonly)
 {
     LOCK(cs_KeyStore);
-    if (!memonly && !WalletBatch(m_storage.GetDatabase()).WriteHDChain(chain))
-        throw std::runtime_error(std::string(__func__) + ": writing chain failed");
+    // memonly == true means we are loading the wallet file
```

Comment:
This is really crying out to be cleaned up:

- there's a bool flag that substantially changes the logic of this function
- the function is called in two places only, once with `true` and once with `false`
- the argument name is confusing enough that it requires an inline comment that explains where the function is being called from.

It seems to me that this function should be split into a `WriteHDChain()` and `SetHDChain()` functions.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/17681#discussion_r427633081

> This is really crying out to be cleaned up:

Agree with this, but my immediate reaction here (without looking too deeply) is that the changes suggested here are only tangentially related to this PR and should be made in a separate PR before or after this one. There's already a lot going on here and it seems like you could make similar comments about a lot of the other wallet functions this PR touches.

> It seems to me that this function should be split into a `WriteHDChain()` and `SetHDChain()` functions.

Again didn't look too closely, but a current convention is to have CWallet and KeyMan LoadXXX and AddXXX methods

Reply:
>  the changes suggested here are only tangentially related to this PR

Prior to this PR, the `memonly` flag only controlled whether the new seed was written to the database. This PR changes things so the majority of the function is in an `if` block that depends on the flag, so I'd argue it's not tangential.

I think it's fine to do this in a follow-up, but if this PR gets retouched, no problem doing it here either.

Reply:
I agree that this should be cleaned up. There are a few other functions that follow this same pattern, so I think it would be better to change them all at the same time in a followup PR.